rust/sbp/src/sbp_iter_ext.rs

```diff
@@ -123,66 +115,93 @@ where
-impl<I> RoverTimeIter<I>
-where
-    I: Iterator,
-{
-    fn new(messages: I) -> RoverTimeIter<I> {
-        Self {
-            messages,
-            clock: None,
+    impl<I> RoverTimeIter<I>
+    where
+        I: Iterator,
+    {
+        pub fn new(messages: I) -> RoverTimeIter<I> {
+            Self {
+                messages,
+                clock: None,
+            }
         }
-    }
 
-    fn update(&mut self, time: MessageTime) -> Option<GpsTime> {
-        match time {
-            MessageTime::Rover(time) => match time {
-                RoverTime::GpsTime(time) => {
-                    self.clock = Some(time);
-                    Some(time)
-                }
-                RoverTime::Tow(tow) => {
-                    if let Some(clock) = self.clock {
-                        // tow came from a `GpsTime` so it must be valid
-                        let time = GpsTime::new(clock.wn(), tow).unwrap();
+        fn update(&mut self, time: MessageTime) -> Option<GpsTime> {
+            match time {
+                MessageTime::Rover(time) => match time {
+                    RoverTime::GpsTime(time) => {
                         self.clock = Some(time);
                         Some(time)
-                    } else {
-                        // no previous time to use wn from
-                        None
                     }
-                }
-            },
-            _ => None,
+                    RoverTime::Tow(tow) => {
+                        if let Some(clock) = self.clock {
+                            // tow came from a `GpsTime` so it must be valid
+                            let time = GpsTime::new(clock.wn(), tow).unwrap();
+                            self.clock = Some(time);
+                            Some(time)
+                        } else {
+                            // no previous time to use wn from
+                            None
+                        }
+                    }
+                },
+                _ => None,
+            }
         }
     }
-}
 
-#[cfg(feature = "swiftnav")]
-impl<I> Iterator for RoverTimeIter<I>
-where
-    I: Iterator<Item = Sbp>,
-{
-    type Item = (Sbp, Option<Result<GpsTime, GpsTimeError>>);
+    impl<I> Iterator for RoverTimeIter<I>
+    where
+        I: Iterator,
+        I::Item: HasTime,
+    {
+        type Item = (I::Item, Option<Result<GpsTime, GpsTimeError>>);
+
+        fn next(&mut self) -> Option<Self::Item> {
+            let msg = self.messages.next()?;
+            match msg.time() {
+                Some(Ok(time)) => match self.update(time) {
+                    Some(gps_time) => Some((msg, Some(Ok(gps_time)))),
+                    None => Some((msg, None)),
+                },
+                Some(Err(e)) => Some((msg, Some(Err(e)))),
+                None => Some((msg, None)),
+            }
+        }
+    }
 
-    fn next(&mut self) -> Option<Self::Item> {
-        let msg = self.messages.next()?;
+    pub trait HasTime {
```

Comment:
Kind of a clunky solution but I don't know a better way to do this without specialization